Kernel/Syscalls/socket.cpp

```diff
@@ -247,24 +248,30 @@ ErrorOr<FlatPtr> Process::sys$recvmsg(int sockfd, Userspace<struct msghdr*> user
     auto description = TRY(open_file_description(sockfd));
     if (!description->is_socket())
         return ENOTSOCK;
-    auto& socket = *description->socket();
 
+    auto& socket = *description->socket();
     if (socket.is_shut_down_for_reading())
         return 0;
 
-    auto data_buffer = TRY(UserOrKernelBuffer::for_user_buffer((u8*)iovs[0].iov_base, iovs[0].iov_len));
     Time timestamp {};
     bool blocking = (flags & MSG_DONTWAIT) ? false : description->is_blocking();
-    auto result = socket.recvfrom(*description, data_buffer, iovs[0].iov_len, flags, user_addr, user_addr_length, timestamp, blocking);
+    int msg_flags = 0;
 
-    if (result.is_error())
-        return result.release_error();
+    Checked<ssize_t> total_length = 0;
+    for (auto const& iov : iovs)
+        total_length += iov.iov_len;
 
-    int msg_flags = 0;
+    if (total_length.has_overflow())
+        return EINVAL;
 
-    if (result.value() > iovs[0].iov_len) {
-        VERIFY(socket.type() != SOCK_STREAM);
-        msg_flags |= MSG_TRUNC;
+    auto data = TRY(ByteBuffer::create_zeroed(total_length.value()));
+    auto data_buffer = UserOrKernelBuffer::for_kernel_buffer(data.data());
+    auto nread = TRY(socket.recvfrom(*description, data_buffer, data.size(), flags, user_addr, user_addr_length, timestamp, blocking));
```

Comment:
Looking at this closer again... what happens if nread < total_length? 

It doesn't look like the loop below will terminate at the end of nread, it will continue to copy zeroes into the user's io vectors until we fill all the buffers? 

Are we supposed to fill the rest of the vectors with 0? or should we leave the user's receive buffers untouched after the end of the data? The VERIFY below looks like it should be accompanied by some kind of "only copy nread bytes total" check.

POSIX (https://pubs.opengroup.org/onlinepubs/9699919799/functions/recvmsg.html) says:

> Each storage area indicated by msg_iov is filled with received data in turn until all of the received data is stored or all of the areas have been filled.

Reply:
It might make sense to make that optimization in a follow-up PR with some tests in Tests/Kernel to check out what happens we actually try to sendmsg/recvmsg with multiple iov attaached. What do you think?

Reply:
Yes the loop will continue to copy zeros if that is the case. Most of the times it is not so I thought it would be ok. I can add a condition to break the loop when the offset is greater than the bytes read if you think that would be a better idea. It would save us some copying to user so it is not a bad idea.

I can add some tests to see how things behave but I'm a bit busy with work so it may take me some time.

Let me know what you think and thanks for the review.

Reply:
@ADKaster, I've added the tests for recvmsg and sendmsg. Also I've changed sendmsg to gather all the messages and send them in one go. This wasn't clear in the posix spec because it was  only described in the Error section.